contracts/src/BoundlessMarket.sol

```diff
@@ -509,8 +516,9 @@ contract BoundlessMarket is
         }
 
         // If no fulfillment context was stored for this request digest (via priceRequest),
-        // then payment cannot be processed. This check also serves as an expiration check since
-        // fulfillment contexts cannot be created for expired requests.
+        // then payment cannot be processed. This check also serves as
+        // 1/ an expiration check since fulfillment contexts cannot be created for expired requests.
+        // 2/ a smart contract signature check, since signatures are validated when a request is priced.
         FulfillmentContext memory context = FulfillmentContextLibrary.load(requestDigest);
         if (!context.valid) {
             return abi.encodeWithSelector(RequestIsNotPriced.selector, RequestId.unwrap(id));
```

Comment:
We need to revert here instead of simply returning a payment error. As written, someone could mark a request as fulfilled without ever passing a signature check.

Reply:
Just want to confirm the case you are describing so I can add a test for it.

Is it:
1. Request priced, where request id indicates ecdsa (so client signature check is skipped during price).
2. Fulfill called for a unlocked request id that indicates smart contract signature, but same request digest (signature check skipped in assessor)
3. Request would be marked as fulfilled and we'd attempt to pay

Reply:
Got it, one interesting thing here is that we're in the wasLocked path, and for the request to have been locked there must have existed a valid SC signature at lock time. Regardless I think it makes sense to confirm the signature is valid again at fulfillment time, as we may need to pull additional funds from the client (in the multiple request same id case).